File: app/api/routes/memory.py
"""Memory API: store and search."""

from uuid import UUID
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user_optional
from app.db.session import get_db
from app.models.memory import Memory, MemoryType
from app.models.user import User
from app.services.memory.store import store_memory
from app.services.memory.retrieval import search_memories

router = APIRouter()


class MemoryCreate(BaseModel):
    content: str = Field(..., min_length=1, max_length=100_000)
    memory_type: str = "note"
    title: Optional[str] = Field(None, max_length=512)


class MemoryOut(BaseModel):
    id: str
    memory_type: str
    content: str
    title: Optional[str]
    created_at: str

    @classmethod
    def from_orm(cls, m: Memory) -> "MemoryOut":
        return cls(
            id=str(m.id),
            memory_type=m.memory_type.value,
            content=m.content,
            title=m.title,
            created_at=m.created_at.isoformat(),
        )


class SearchResult(BaseModel):
    memory: MemoryOut
    chunk_text: str
    score: float


@router.post("/", response_model=MemoryOut)
async def create_memory(
    body: MemoryCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User | None = Depends(get_current_user_optional),
) -> MemoryOut:
    try:
        mt = MemoryType(body.memory_type)
    except ValueError:
        raise HTTPException(400, f"Invalid memory_type. Must be one of: {[e.value for e in MemoryType]}")
    memory = await store_memory(
        db,
        content=body.content,
        memory_type=mt,
        title=body.title,
        user_id=current_user.id if current_user else None,
    )
    return MemoryOut.from_orm(memory)


@router.get("/search", response_model=list[SearchResult])
async def search(
    q: str,
    limit: int = 10,
    db: AsyncSession = Depends(get_db),
    current_user: User | None = Depends(get_current_user_optional),
) -> list[SearchResult]:
    if not q.strip():
        return []
    results = await search_memories(
        db, q, limit=min(limit, 50),
        user_id=current_user.id if current_user else None,
    )
    return [
        SearchResult(memory=MemoryOut.from_orm(m), chunk_text=chunk, score=score)
        for m, chunk, score in results
    ]

File: server/routes/memory.js
import { Router } from "express";
import { storeMemory, searchMemories } from "../services/memory.js";

const VALID_TYPES = ["conversation", "note", "idea", "document", "project", "belief", "goal"];
const router = Router();

router.post("/", async (req, res) => {
  try {
    const { content, memory_type = "note", title } = req.body || {};
    if (!content || typeof content !== "string") {
      return res.status(422).json({ detail: "content required" });
    }
    if (content.length > 100_000) {
      return res.status(422).json({ detail: "content too long" });
    }
    const mt = String(memory_type || "note").toLowerCase();
    if (!VALID_TYPES.includes(mt)) {
      return res.status(400).json({
        detail: `Invalid memory_type. Must be one of: ${VALID_TYPES.join(", ")}`,
      });
    }
    const memory = await storeMemory(content, mt, title || null, req.userId);
    res.json({
      id: memory.id,
      memory_type: memory.memory_type,
      content: memory.content,
      title: memory.title,
      created_at: memory.created_at,
    });
  } catch (err) {
    console.error(err);
    res.status(500).json({ detail: "Internal server error" });
  }
});

router.get("/search", async (req, res) => {
  try {
    const q = String(req.query.q || "").trim();
    const limit = Math.min(parseInt(req.query.limit || "10", 10) || 10, 50);
    if (!q) return res.json([]);
    const results = await searchMemories(q, limit, req.userId);
    res.json(
      results.map((row) => ({
        memory: {
          id: row.id,
          memory_type: row.memory_type,
          content: row.content,
          title: row.title,
          created_at: row.created_at,
        },
        chunk_text: row.chunk_text,
        score: parseFloat(row.score),
      })
    );
  } catch (err) {
    console.error(err);
    res.status(500).json({ detail: "Internal server error" });
  }
});

export default router;
